paginator: manual pagination with familiar syntax (#337)
get-convex/convex-helpers: packages/convex-helpers/README.md

});
```

### `paginator`: manual pagination with familiar syntax

> In addition to `getPage`, convex-helpers provides a function
> `paginator`. This function has syntax and interface similar to the
> built-in Convex query builder, to make it easy to switch.
> 
> It runs on top of `getPage`, so it provides the benefits of being callable
> multiple times from a query, or within a Convex component. On the other hand,
> it provides less control over the index ranges being queried.
> 
> The interface is so similar to `.paginate` that you can use it with
> `usePaginatedQuery`. **However**, doing so will result in non-reactive pages.
> Keeping pages contiguous requires rerunning queries and passing through an
> `endCursor` option. For more info on these edge-cases, see
> https://stack.convex.dev/fully-reactive-pagination.
> 
> `paginator` is especially useful when you cannot use the built-in `.paginate`
> and you are not in a reactive query. For example, if you're running a migration,
> it's running in a mutation and it doesn't need reactivity. You can run
> multiple migrations at once or run a migration within a Convex component using
> `paginator`.

Reviewer: ```suggestion
In addition to `getPage`, convex-helpers provides a function
`paginator` as an alternative to the built-in `db.query.paginate`.

- The built-in `.paginate` is currently limited to one call per query, which allows
  it to track the page's "end cursor" for contiguous reactive pagination client-side.
- `paginator` can be called multiple times from a query,
  but does not subscribe the query to the end cursor automatically.

The syntax and interface for `paginator` is so similar to `.paginate` that it is
nearly a drop-in replacement and can even be used with `usePaginatedQuery`.
This makes it more suitable for non-reactive pagination usecases,
such as iterating data in a mutation. Note: it supports `withIndex` but not `filter`.

For more information on reactive pagination and end cursors, see
https://stack.convex.dev/fully-reactive-pagination
and
https://stack.convex.dev/pagination

**However**, the pages may not stay contiguous.
```

> As a basic example, suppose you have this query:
> 
> ```ts
> export const list = query({
>   args: { opts: paginationOptsValidator },
>   handler: async (ctx, { opts }) => {
>     return await ctx.db.query("messages").paginate(opts);
>   },
> });
> ```
> 
> It has the same behavior as this query, except that in this one the pages might
> not stay contiguous as items are added and removed from the list and the query
> updates reactively:
> 
> ```ts
> import { paginator } from "convex-helpers/server/pagination";
> import schema from "./schema";
> 
> export const list = query({
>   args: { opts: paginationOptsValidator },
>   handler: async (ctx, { opts }) => {
>     return await paginator(ctx.db, schema).query("messages").paginate(opts);
>   },
> });
> ```

Reviewer: ```suggestion
As a basic example, consider replacing this query with `paginator`.
It has the same behavior, except that the pages might not stay contiguous as
items are added and removed from the list and the query updates reactively.

```ts
import { paginator } from "convex-helpers/server/pagination";
import schema from "./schema";

export const list = query({
  args: { opts: paginationOptsValidator },
  handler: async (ctx, { opts }) => {
    // BEFORE:
    return await ctx.db.query("messages").paginate(opts);
    // AFTER:
    return await paginator(ctx.db, schema).query("messages").paginate(opts);
  },
});
```
```
